File: application-server/node/src/controllers/token.controller.js
import { Router } from "express";
import { LIVEKIT_API_KEY, LIVEKIT_API_SECRET } from "../config.js";
import { AccessToken } from "livekit-server-sdk";

export const tokenController = Router();

tokenController.post("/", async (req, res) => {
    const { roomName, participantName } = req.body;

    if (!roomName || !participantName) {
        res.status(400).json({ errorMessage: "'roomName' and 'participantName' are required" });
        return;
    }

    const at = new AccessToken(LIVEKIT_API_KEY, LIVEKIT_API_SECRET, {
        identity: participantName
    });

    const grant = {
        roomJoin: true,
        room: roomName,
        canPublish: true,
        canSubscribe: true
    };
    at.addGrant(grant);

    const token = await at.toJwt();
    res.json({ token });
});

File: application-server/go/controllers/token_controller.go
package controllers

import (
	"net/http"
	"openvidu/go/config"

	"github.com/gin-gonic/gin"
	"github.com/livekit/protocol/auth"
)

func TokenRoutes(router *gin.Engine) {
	router.POST("/token", createToken)
}

func createToken(c *gin.Context) {
	var body struct {
		RoomName        string `json:"roomName" binding:"required"`
		ParticipantName string `json:"participantName" binding:"required"`
	}

	if err := c.ShouldBindJSON(&body); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"errorMessage": "'roomName' and 'participantName' are required"})
		return
	}

	at := auth.NewAccessToken(config.LivekitApiKey, config.LivekitApiSecret)
	grant := &auth.VideoGrant{
		RoomJoin: true,
		Room:     body.RoomName,
	}
	at.SetVideoGrant(grant).SetIdentity(body.ParticipantName)

	token, err := at.ToJWT()
	if err != nil {
		c.JSON(http.StatusInternalServerError, err.Error())
		return
	}

	c.JSON(http.StatusOK, gin.H{"token": token})
}
